sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -269,4 +323,214 @@ case class InsertIntoHadoopFsRelationCommand(
+         | ($outputPath).
+         | Detailed information (conflicted path , appId, last modification time):
+         | $relativePath,$appId,$modificationTime.
+         |
+         | There could be two possibilities:
+         | 1. The path is being written by another InsertDataSource operation and you need wait
+         |    for it to complete.
+         | 2. This path belongs to a failed application who didn't have a chance to clean it up
+         |    gracefully.
+         |
+         | Please check with the provided appId and last modification time
+         | whether another application is still writing to the path .
+         | If not, it's safe to delete it and retry your application.
+         |""".stripMargin)
+  }
+
+  /**
+   * Find relative staging partition paths, which is conflicted with current
+   * InsertIntoHadoopFsRelation operation.
+   */
+  private def findConflictedStagingPartitionPaths(
+      fs: FileSystem,
+      path: Path,
+      depth: Int): Seq[Path] = {
+    val paths = ListBuffer[Path]()
+    try {
+      if (fs.exists(path)) {
+        if (depth == 0) {
+          paths += path
+        } else {
+          for (file <- fs.listStatus(path)) {
+            paths ++= findConflictedStagingPartitionPaths(fs, file.getPath, depth - 1)
+          }
+        }
+      }
+    } catch {
+      case e: Exception =>
+        logError("Exception occurred when finding conflicted staging partition paths.")
+        throw e
+    }
+    paths
+  }
+
+  /**
+   * Try to get the lock of a partition. If the fileSystem is not [[DistributedFileSystem]], check
+   * whether the lock file exists, if existed, we can not get the partition lock, otherwise, create
+   * the lock file and get the partition lock.
+   * Otherwise, this filesystem is [[DistributedFileSystem]], if the lock file exists, we can try to
+   * append this lock file, if append successfully, we get the partition lock; else if the lock file
+   * does not exist, we create this lock file and get the partition lock.
+   */
+  private def tryLock(
+      stagingPartitionPath: Path,
+      fs: FileSystem): Boolean = {
+    try {
+      if (!fs.exists(stagingPartitionPath)) {
+        fs.mkdirs(stagingPartitionPath)
+      }
+      val partitionLock = new Path(stagingPartitionPath, getLockName)
+      if (!fs.isInstanceOf[DistributedFileSystem]) {
+        val existed = fs.exists(partitionLock)
+        if (!existed) {
+          fs.create(partitionLock).close()
+        }
+        !existed
+      } else {
+
+        def pickLock: Boolean = Try {
+          if (fs.exists(partitionLock)) {
+            fs.append(partitionLock)
+          } else {
+            fs.create(partitionLock)
+          }
+        } match {
+          case Success(_) => true
+          case _ => false
+        }
+
+        if (pickLock) {
+          true
+        } else {
+          Thread.sleep(HdfsConstants.LEASE_SOFTLIMIT_PERIOD)
```

Comment:
This is quite some time(IIRC, 60secs), should we add an option to turn off this wait?

Reply:
Here we just double check after a soft limit period when the first check is failed.

In fact, we can set an timeout and try to pick lock with a period, such as 1000 ms.